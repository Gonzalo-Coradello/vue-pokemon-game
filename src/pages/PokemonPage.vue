<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <div class="score">
            <p>Puntaje actual: {{ currentScore }}</p>
            <p>Puntaje máximo: {{ highestScore }}</p>
        </div>
        <h1>¿Quién es este Pokémon?</h1>
        <PokemonPicture 
            :pokemon-id="pokemon.id" 
            :show-pokemon="showPokemon" 
        />
        <PokemonOptions 
            :pokemons="pokemonArr" 
            @selection="checkAnswer" 
        />

        <template v-if="showAnswer" class="fade-in">
            <h2>{{ message }}</h2>
            <h3 v-if="currentScore === 0">¡Acertaste {{ lastScore }}!</h3>
            <button @click="newGame">{{ currentScore === 0 ? 'Nuevo juego' : 'Siguiente' }}</button>
        </template>
    </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: { PokemonPicture, PokemonOptions },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
            currentScore: 0,
            lastScore: 0,
            highestScore: 0
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[rndInt]
        },
        checkAnswer(selectedId) {
            this.showPokemon = true
            this.showAnswer = true

            if(selectedId === this.pokemon.id) {
                this.message = `Correcto, ${this.pokemon.name}`
                this.currentScore++
                if (this.currentScore > this.highestScore) {
                    this.highestScore = this.currentScore
                }
            } else {
                this.message = `Oops, era ${this.pokemon.name}`
                this.lastScore = this.currentScore
                this.currentScore = 0
            }
        },
        newGame() {
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null

            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}

</script>

<style scoped>
.score {
    text-align: right;
    margin-bottom: 1rem;
}
</style>