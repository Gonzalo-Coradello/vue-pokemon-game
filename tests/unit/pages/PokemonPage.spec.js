import { describe, test, expect, beforeEach, vi } from 'vitest'
import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {
    
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })
    
    test('debe llamar mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = vi.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('debe hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()  
    })

    test('debe mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')
        // const picture = wrapper.findComponent(PokemonPicture)
        // const options = wrapper.findComponent(PokemonOptions)

        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        expect(picture.attributes('pokemonid')).toBe('1')
        expect(options.attributes('pokemons')).toBeTruthy
    })

    test('pruebas con checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(10)
        
        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)
    })
})