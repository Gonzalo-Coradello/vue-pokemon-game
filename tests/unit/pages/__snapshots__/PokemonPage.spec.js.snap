// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonPage Component > debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
"<div data-v-2108564c="">
  <div data-v-2108564c="" class="score">
    <p data-v-2108564c="">Puntaje actual: 0</p>
    <p data-v-2108564c="">Puntaje máximo: 0</p>
  </div>
  <h1 data-v-2108564c="">¿Quién es este Pokémon?</h1>
  <pokemon-picture-stub data-v-2108564c="" pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub data-v-2108564c="" pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>"
`;
